### DISCUSSION TITLE
Network connections like `ssh` within `tmux` are broken after updating Ghostty

### DISCUSSION THREAD

**Discussion by fooness** [ghostty-org/ghostty#2998] [Q&A] [answered]
This is a weird one, and I’m not sure I can describe it rather good, but I can reproduce the problem.

1. Start Ghostty
2. Run `tmux` to start new session
3. Run e.g. `ssh <remote>` -> connection established
4. Check for Ghostty update
5. Update/Restart Ghostty
6. Run `tmux attach`
7. SSH connection was closed
8. Run e.g. `ssh <remote>` -> "No route to host"
10. Run `ping <remote>` -> this does work, though
11. Quit `tmux` session
12. Run `tmux` to start new session
13. Run e.g. `ssh <remote>` -> connection established

I am on macOS Sequoia, my macOS Firewall is enabled, and I run Little Snitch; but neither the enabled macOS Firewall nor running Little Snitch should matter here, as everything works as expected until updating Ghostty, and work again after quitting and starting `tmux` again.

**Comment by mitchellh** [comment]
This is weird enough that I'm going to make it a Q&A until we get some additional confirmation here. We have a LOT of tmux users and a lot of infrastructure folks in the beta so I would've expected to hear something like this before. So I want to just clear it up in discussions until we convert to an issue.

**Reply by fooness** [reply]
Update: For me, the issue still persists through `v1.0.0` and `v1.0.1` …

Unfortunately though, I myself am not really sure on how to inspect or debug this issue any further.

**Comment by hvdute** [comment]
I'm using ghostty version `a2445359` on macos 15.2 and this problem still happens. It happens randomly since i started using ghostty. Mostly when connecting to services running on LAN. For example:
<img width="1248" alt="image" src="https://github.com/user-attachments/assets/76660b86-1c85-4c91-8d98-44c9f5fd16fc" />

It works when detached from tmux session. After running `tmux kill-server` and start tmux over again, it works.
<img width="843" alt="image" src="https://github.com/user-attachments/assets/9f1d2206-587b-4ef7-a287-a1e011190f23" />

**Comment by kassadin** [comment]
```ping 192.168.31.167
PING 192.168.31.167 (192.168.31.167): 56 data bytes
ping: sendto: No route to host
ping: sendto: No route to host
Request timeout for icmp_seq 0
ping: sendto: No route to host
Request timeout for icmp_seq 1
ping: sendto: No route to host
```
I tested in Terminal, WezTerm, and iTerm2, and only Ghostty didn't work.
However, it can ping the gateway(192.168.31.1)

**Comment by mybigman** [comment]
This should be flagged as a bug not a discussion.

This is definitely an issue... I cant ping or ssh to anything on my local network using ghostty. 

Everything works in terminal or iterm2.

```
ghostty --version
Ghostty 1.0.1

Version
  - version: 1.0.1
  - channel: stable
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.none
  - font engine: font.main.Backend.coretext
  - renderer   : renderer.Metal
  - libxev     : main.Backend.kqueue
 ```
 
 ```
 macOS Sequoia 15.2
```

```
> ssh username@192.168.1.71
ssh: connect to host 192.168.1.71 port 22: No route to host

> ping 192.168.1.71
PING 192.168.1.71 (192.168.1.71): 56 data bytes
ping: sendto: No route to host
ping: sendto: No route to host
Request timeout for icmp_seq 0

> nmap -sS -O -p22 192.168.1.71
Starting Nmap 7.95 ( https://nmap.org ) at 2025-01-30 19:22 AEDT
Nmap scan report for 192.168.1.71
Host is up (0.017s latency).

PORT   STATE SERVICE
22/tcp open  ssh
```

**Reply by rhodes-b** [reply]
> This should be flagged as a bug not a discussion

> [this](https://github.com/ghostty-org/ghostty/discussions/2998#discussioncomment-11594486) -> This is weird enough that I'm going to make it a Q&A until we get some additional confirmation here. We have a LOT of tmux users and a lot of infrastructure folks in the beta so I would've expected to hear something like this before. So I want to just clear it up in discussions until we convert to an issue.

Nobody said this isnt a bug. It seens only very small number of users having are having this issue.  Things get converted into an issue (github issue) when its actionable (reproducible) until then it stays as a discussion to figure out what is going on

**Comment by fooness** [comment]
I just updated to version `1.1.0`, re-attached my `tmux` session, and the problem still persists; I had to kill my `tmux` session and start all over, and then `ssh`, `ping`, etc. started working again.

**Comment by mybigman** [comment]
ive upgraded to 1.1.0 still same issue persists

**Comment by taylrfnt** [comment]
Been doing some digging on a similar issue report that came through the Discord... not sure if this is universal or limited to MacOS, but the majority of the reports below appear to be limited to MacOS, so sharing what appears to work for most cases.  I have not been able to reproduce the issues within tmux, but anything outside a multiplexer should work with the fix below:

Apple has added some "enhanced security" features across various macOS versions that (sadly) most users tend to decline/dismiss, since it is not well-documented.  In particular, on fresh installs, Ghostty (or any app that uses select network tools/features, for that matter) will request access to access devices on your local network.  If you decline/dismiss, it kneecaps Ghostty's access to the network, including things like ping, telnet, etc.  

You can use [Apple's documentation](https://support.apple.com/guide/mac-help/control-access-to-your-local-network-on-mac-mchla4f49138/mac) to allow Ghostty permissions on your local network, fully quit (`super`+`q`) Ghostty, then relaunch and it should work.  Apple unfortunately does not have a good document on the initial alert itself, and I can't reliably get it to appear, so I'll add a screenshot the next time I encounter it as a reply to this comment.

**Reply by mybigman** [reply]
This was it. I dont even recall seeing a popup and I only recently switched to ghostty from iterm2.

It is resolved for me now. thanks.

**Reply by fooness** [reply]
That’s not the issue for me (and some other commenters here, according to their comments/reports); the permissions are correct, the networking works inside Ghostty without a terminal multiplexer like `tmux`, but not within `tmux` after an update.

**Reply by edrozenberg** [reply]
> ...but not within `tmux` after an update.

Are you using ssh agent forwarding by any chance?

Found several older references to tmux issues related to `SSH_AUTH_SOCK`. But not sure it's even relevant any more with recent tmux versions.

Tmux and SSH agent forwarding
without-brains.net/2020/08/05/tmux-and-ssh-agent-forwarding/

tmux: Cannot SSH from inside tmux, even though I have agent-forwarding (ssh -A) on [duplicate]
superuser.com/questions/334975/tmux-cannot-ssh-from-inside-tmux-even-though-i-have-agent-forwarding-ssh-a

tmux cannot work with ssh-agent
stackoverflow.com/questions/49382507/tmux-cannot-work-with-ssh-agent

**Reply by fooness** [reply]
> Are you using ssh agent forwarding by any chance?

No.

**Reply by mrbuk** [reply]
I had a similar issue but IMO this was related to `tmux` itself and how it behaves with the new Sequoia [Local Network Privacy](https://developer.apple.com/forums/thread/663858) feature.

I typically observed the behaviour after 
1. running `tmux` for some time including couple of laptop sleep states or 
2. after the instance of Terminal / ghostty was terminated but the `tmux` server continued running.

So my assumption is that `tmux` somehow gets attached / associated to the [Local Network Privacy](https://developer.apple.com/forums/thread/663858) of the process that is used to start the tmux server.

To address this I now start tmux using [launchd](https://gist.github.com/mrbuk/4244901383265997f4af5b28cf627520). After doing this the first time I ran a `ssh` in a newly created `tmux` session and got the following pop-up asking for permissions. 

So far I have not run into the issue again.

<img width="258" alt="image" src="https://github.com/user-attachments/assets/dc412caf-a14b-4676-99a0-148862e564a9" />

**Reply by fooness** [reply]
The issue might indeed be related to the [Local Network Privacy](https://developer.apple.com/forums/thread/663858) feature, but it’s kinda weird that this problem happens in Ghostty, but not in e.g. iTerm2 or Terminal.app …

**Reply by jasondinh** [reply]
@fooness it happened with Terminal.app for me. I'm doing the launchd thing and tmux is now properly allowed to access local network, hopefully it won't break again. If it does I will come back and update this comment.

**Reply by jlehtimaki** [reply]
For me it got broken, again.

**Reply by gnp-x** [reply]
This is what fixed my issue. Strangely enough, I was able to ssh into an older machine just fine with local network setting off for ghostty, but when I added a new machine it wasn't working. Giving it local network access fixed my problem.

**Comment by afonsofrancof** [comment]
I just spent 30 minutes trying to understand why I couldn't ssh to my Raspberry Pi from my M3 MacBook Pro.
Pinging the host worked, and using netcat on port 22 worked, and I could see ssh traffic on Wireshark, but not when using the ssh command directly. I kept getting "no route to host".

Somehow my tmux instance got borked and now this error shows up regardless of which terminal emulator I use.
I don't know if ghostty broke my tmux session, but I am leaving this info here, hoping it can help.

**Reply by taylrfnt** [reply]
Did you try the solution recommended just a few posts above yours?  If so, did it work?
https://github.com/ghostty-org/ghostty/discussions/2998#discussioncomment-12109408

**Reply by afonsofrancof** [reply]
I never got the prompt asking for permission to access my LAN.
I checked in the settings and tmux is not listed there so I can't toggle it on. There is also no way to add apps manually in the settings.
I will try to use launchd to launch it and see if it asks for permission.
Thanks!

**Reply by edrozenberg** [reply]
Something else to try is to navigate to where the binary is using Finder or other file mgr, then right click -> *Open* on the binary. This can help trigger whatever MacOS permissions are involved, for ex. it can help trigger the item to show up in *System Settings -> Privacy & Security -> Allow to Open* for the binary item.

**Comment by shscy** [comment]
you might need to grant local network access to Ghostty in System Settings -> Privacy & Security -> Local Network.

**Reply by fooness** [reply]
Yes, but that’s unrelated to the problem description.

**Comment by talkingtab** [comment]
I have an issue, not using ghostty, but relevant. I think. 
- Start iterm2 or terminal. 
- start tmux
- attempt to run postgres-node to connect to another host on the local network running a postgres server
- error: HOSTUNREACH
- exit tmux
- run same node program. It works
Again this is true for both terminal and iterm2. I suspect that it is related to how Apple (sigh) has implemented the local network privacy thing. What is supposed to happen is that when a program attempts to access a local network, an entry shows in the privacy/local network thing. But it does not with tmux. Yet the code to test whether access should be given can determine it is tmux not terminal etc. This is stupid.

**Comment by bazuchan** [comment]
JFYI I am no longer having this problem after upgrading to tahoe

**Reply by SamuelBolduc** [reply]
Still having this problem on Tahoe, using Zellij rather than tmux. But it's the same exact symptoms.

**Reply by chrisbennight** [reply]
still having the problem; codesigning tmux seemed to fix it for me:  codesign --force --deep --sign - /opt/homebrew/bin/tmux

**Comment by aSapien** [comment]
Wow I never suspected anything Ghostty+Tmux related. 
Just spent a few hours googling for Nix+Darwin SSH issues as I was quite sure it was the culprit.
To my surprise it suddenly works in a new Ghostty window (before I start Tmux). 🫨 

So, no localnet SSH from Ghostty+Tmux on MacOS?

- MacOS: `Sequoia 15.7.1`
- Ghostty: `1.3.0-main+dffa4f4fc`
- Tmux: `3.5a`

**Reply by talkingtab** [reply]
I suspect that since ghostty and iterm2 have the exact same problem *only* when using a tmux session, that it is an Apple problem. That the code that determines the program trying to access local net finds "tmux", but the code that _allows you to give access_ is not the same. It only finds "iterm2" or "ghostty" and does not look deeper. Someone posted something saying this was recently fixed?

**Comment by mitchellh** [comment, marked as answer]
There is clearly some issue here. As @talkingtab noted (and I agree), this doesn't seem to be a Ghostty problem since it is reproducible in multiple terminals. Given this has been so long and we haven't reached an actionable, reproducible bug that we can work on, I'm going to close this. If someone can produce something that we can use to convert to an issue, please open a new discussion.

**Comment by spm1001** [comment]
So... long time but I found this post yesterday and I'm writing to offer my MacBook as a study subject in case it helps narrow things down since I'm seeing similar behaviour:

- In Ghostty (macOS 27.0 beta, build 26A5388g, Ghostty 1.3.1, bundle 15212, stable channel) I get the output below. But in Terminal.app the second ping completes correctly. 

```
launchctl submit -l t1 -o /tmp/p.out -e /tmp/p.out -- /sbin/ping -c 2 -W 1000 <lan-peer>
  → 2 packets received, at uid=501

/sbin/ping -c 2 -W 1000 <lan-peer>          # same shell, inside Ghostty
  → ping: sendto: No route to host  (0 received)
```

And in the logs I see an error for LocalNetwork being blocked for Ghostty on exactly the failing pings, and nothing else:

```
/usr/bin/log stream --predicate 'subsystem == "com.apple.TCC" OR process == "tccd" OR eventMessage CONTAINS[c] "local network"'

UserEventAgent (com.apple.networkextension): Got local network blocked notification: pid N
UserEventAgent [com.apple.networkextension]: LocalNetwork: found bundle id com.mitchellh.ghostty by PID
UserEventAgent [com.apple.networkextension]: LocalNetwork: ... by UUID
```

Ghostty doesn't appear in my list of apps under Privacy&Security > Local Network, so I can't grant it manually and in `/Library/Preferences/com.apple.networkextension.plist` Ghostty carries `MulticastPreferenceSet=false / DenyMulticast=true` which I think is the never asked, default deny. 

Can I do anything else to help identify what's happening?

**Reply by jasondinh** [reply]
This seems to be a macOS 27 beta issue. I'm on the same build as you, I updated my Firefox and couldn't get macOS to prompt for LAN access again.

**Reply by BTForIT** [reply]
> This seems to be a macOS 27 beta issue. I'm on the same build as you, I updated my Firefox and couldn't get macOS to prompt for LAN access again.

this my current issue, did you figure it out?

**Reply by jasondinh** [reply]
@BTForIT no I assume it's a macOS issue and will wait for the next beta. Can't expect much since it's still in beta after all.

**Reply by noon-ng** [reply]
Also on Golden Gate and having this issue, although interestingly nothing concerning `com.mitchellh.ghostty` shows up in the log when I try pinging my router address.

I dunno if this helps, but if I start Ghostty by calling the binary directly from Terminal.app, I can ping LAN addresses just fine from that Ghostty window.

**Reply by jasondinh** [reply]
Beta 5 fixed this for me.
